proposals/lip-Introduce_PoA_module.md

```diff
@@ -0,0 +1,607 @@
+Then processing `trs` has the following effect:
+
+*   The array `snapshotStore(2).validators` is set to `trs.params.newValidators`.
+*   The property `snapshotStore(2).threshold` is set to `trs.params.threshold`.
+*   The property `chainProperties.validatorsUpdateNonce` is set to `trs.params.validatorsUpdateNonce + 1`.
+
+### Internal Function
+
+#### shuffleValidatorsList
+
+A function to reorder the list of validators as specified in [LIP 0003][lip3].
+
+##### Parameters
+The function has the following input parameters in the order given below:
+
+* `validatorsList`: An array of pairwise distinct 20-byte addresses.
+* `randomSeed`: A 32-byte value representing a random seed.
+
+##### Returns
+This function returns an array of bytes with the re-ordered list of addresses.
+
+##### Execution
+The shuffling algorithm is defined in [LIP 0003][lip3].
+
+### Protocol Logic for Other Modules
+
+This module does not define any specific logic for other modules.
+
+### Endpoints for Off-Chain Services
+
+This module does not define any endpoint for off-chain services.
+
+### Protocol Logic During Block Lifecycle
+
+#### After Genesis Block Execution
+
+After the genesis block `g` is executed, the following logic is executed:
+
+```python
+chainProperties.roundEndHeight = g.header.height + len(snapshotStore(0).addresses)
+
+# Pass the required information to the BFT module
+BFTThreshold = snapshotStore(0).threshold
+setBFTParameters(BFTThreshold, BFTThreshold, snapshotStore(0).validators)
+
+# Pass the list of validators to the Validators module
+for i in range(len(snapshotStore(0).validators)):
+    addresses =  snapshotStore(0).address[i]
+    
+setGeneratorList(addresses)
+```
+
+where:
+
+* `setBFTParameters` is a function exposed by the [BFT module][BFTmodule].
+* `setGeneratorList`is a function exposed by the [Validators module][validatorsModule].
+
+#### After Block Execution
+
+After a block `b` is executed, the following logic is executed: 
+
+```python
+if b.header.height == chainProperties.roundEndHeight
+    
+    # get the last stored BFT parameters, and update them if needed
+    BFTThreshold = snapshotStore(0).threshold
+    currentBFTParameters = getBFTParameters(b.header.height)
+    if (currentBFTParameters.validators != snapshotStore(0).validators
+        or currentBFTParameters.precommitThreshold != BFT_THRESHOLD
+        or currentBFTParameters.certificateThreshold != BFT_THRESHOLD):
+        # Pass the required information to the BFT module
+        setBFTParameters(BFTThreshold, BFTThreshold, snapshotStore(0).validators)
+    
+    # Reshuffle the list of validators and pass it to the Validators module
+    roundStartHeight = chainProperties.roundEndHeight - len(snapshotStore(0).validators) + 1
+    randomSeed = getRandomBytes(roundStartHeight, len(snapshotStore(0).validators))
+    
+    for i in range(len(snapshotStore(0).validators)):
+    addresses =  snapshotStore(0).address[i]
```

Comment:
The indentation here is incorrect. It would also check if tabs and spaces are mixed here.

Reply:
fixed